File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
        
def solution(maps):
    dx = [0, 0, 1, -1]
    dy = [-1, 1, 0, 0]
    n = len(maps)
    m = len(maps[0])
    def bfs(x, y):
        queue = deque()
        queue.append((x, y))
        while queue:
            x, y = queue.popleft()
            for i in range(4):
                nx = x + dx[i]
                ny = y + dy[i]
                if 0<=nx<n and 0<=ny<m and maps[nx][ny] == 1:
                    maps[nx][ny] = maps[x][y] + 1
                    queue.append((nx, ny))
        
    bfs(0, 0)
    if maps[n-1][m-1] > 1:
        return maps[n-1][m-1] 
    else:
        return -1

File: 프로그래머스/lv2/1844. 게임 맵 최단거리/게임 맵 최단거리.java
import java.util.*;
class Solution {
    static int answer;
    public void bfs(int[][] maps, int r, int c){
        int[] dr = {1,-1, 0, 0};
        int[] dc = {0, 0, -1, 1};
        
        Queue<int[]> queue = new LinkedList<>();
        queue.offer(new int[]{r, c});
        maps[r][c] = 1;
        
        while(!queue.isEmpty()){
            int[] poll = queue.poll();
            r = poll[0];
            c = poll[1];
            for(int i=0; i<4; i++){
                int nr = r + dr[i];
                int nc = c + dc[i];
                if(nr>=0 && nr<maps.length && nc>=0 && nc<maps[0].length && maps[nr][nc] == 1){
                    maps[nr][nc] = maps[r][c] + 1;
                    queue.offer(new int[]{nr, nc});
                }
            }
        }
        
    }
    public int solution(int[][] maps) {
        int r = maps.length - 1;
        int c = maps[0].length - 1;
        
        bfs(maps, 0, 0);
        
        return maps[r][c] == 1 ? -1 : maps[r][c];
    }
}